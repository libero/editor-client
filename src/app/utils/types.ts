import React from 'react';

export type ReactFCProps<FC> = FC extends React.FC<infer P> ? P : {};

export type ValueOf<T> = T[keyof T];
