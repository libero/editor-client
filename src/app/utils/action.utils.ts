import { createAction, SimpleActionCreator } from 'redux-act';

export interface Action<T> {
  type: string;
  payload: T;
}

export interface AsyncAction<R, T> {
  request: SimpleActionCreator<R>;
  success: SimpleActionCreator<T>;
  error: SimpleActionCreator<Error>;
}

export function createAsyncAction<R, T>(type: string): AsyncAction<R, T> {
  return {
    request: createAction<R>(type + '_REQUEST'),
    success: createAction<T>(type + '_SUCCESS'),
    error: createAction<Error>(type + '_ERROR')
  } as AsyncAction<R, T>;
}
