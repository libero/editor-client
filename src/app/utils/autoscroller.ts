const WINDOW_SCROLL_ZONE = 300;

interface Coords {
  x: number;
  y: number;
}

interface ScrollUpdateInterface {
  translate: Coords;
  maxTranslate: Coords;
  minTranslate: Coords;
  minScroll: Coords;
  width: number;
  height: number;
}

export class AutoScroller {
  private static instance: AutoScroller;

  public static getInstance(): AutoScroller {
    if (!AutoScroller.instance) {
      AutoScroller.instance = new AutoScroller();
    }
    return AutoScroller.instance;
  }

  private interval: number | null;
  private container: Element;
  private isAutoScrolling: boolean;

  constructor() {
    this.container = window.document.scrollingElement || window.document.documentElement;
  }

  public clear(): void {
    if (this.interval == null) {
      return;
    }

    clearInterval(this.interval);
    this.interval = null;
  }

  public updateScroll(event: DragEvent, dropContainer: Element): void {
    const containerRect = this.container.getBoundingClientRect();
    const dropRect = dropContainer.getBoundingClientRect();

    const minTranslate: Coords = { x: 0, y: 0 };
    const maxTranslate: Coords = {
      x: containerRect.width,
      y: containerRect.height
    };

    const minScroll = {
      x: dropRect.left + this.container.scrollLeft,
      y: dropRect.top + this.container.scrollTop
    };

    this.doAutoscroll({
      translate: { x: event.clientX, y: event.clientY },
      minTranslate,
      maxTranslate,
      minScroll,
      width: WINDOW_SCROLL_ZONE * 2,
      height: WINDOW_SCROLL_ZONE * 2
    });
  }

  private doAutoscroll({ translate, minTranslate, maxTranslate, width, height }: ScrollUpdateInterface): void {
    const direction = {
      x: 0,
      y: 0
    };

    const speed = {
      x: 1,
      y: 1
    };

    const acceleration = {
      x: 10,
      y: 10
    };

    const { scrollTop, scrollLeft, scrollHeight, scrollWidth, clientHeight, clientWidth } = this.container;

    const isTop = scrollTop === 0;
    const isBottom = scrollHeight - scrollTop - clientHeight === 0;
    const isLeft = scrollLeft === 0;
    const isRight = scrollWidth - scrollLeft - clientWidth === 0;

    if (translate.y >= maxTranslate.y - height / 2 && !isBottom) {
      // Scroll Down
      direction.y = 1;
      speed.y = acceleration.y * Math.abs((maxTranslate.y - height / 2 - translate.y) / height);
    } else if (translate.x >= maxTranslate.x - width / 2 && !isRight) {
      // Scroll Right
      direction.x = 1;
      speed.x = acceleration.x * Math.abs((maxTranslate.x - width / 2 - translate.x) / width);
    } else if (translate.y <= minTranslate.y + height / 2 && !isTop) {
      // Scroll Up
      direction.y = -1;
      speed.y = acceleration.y * Math.abs((translate.y - height / 2 - minTranslate.y) / height);
    } else if (translate.x <= minTranslate.x + width / 2 && !isLeft) {
      // Scroll Left
      direction.x = -1;
      speed.x = acceleration.x * Math.abs((translate.x - width / 2 - minTranslate.x) / width);
    }

    if (this.interval) {
      this.clear();
      this.isAutoScrolling = false;
    }

    if (direction.x !== 0 || direction.y !== 0) {
      this.interval = window.setInterval(() => {
        this.isAutoScrolling = true;
        const offset = {
          left: speed.x * direction.x,
          top: speed.y * direction.y
        };
        this.container.scrollTop += offset.top;
        this.container.scrollLeft += offset.left;
      }, 5);
    }
  }
}
