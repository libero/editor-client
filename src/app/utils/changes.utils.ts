import { cloneDeepWith } from 'lodash';
import { EditorState } from 'prosemirror-state';

import { JSONObject } from 'app/types/utility.types';
import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { BatchChange } from 'app/utils/history/batch-change';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';
import { RearrangingChange } from 'app/utils/history/rearranging-change';
import { UpdateObjectChange } from 'app/utils/history/update-object-change';
import { DeleteObjectChange } from 'app/utils/history/delete-object-change';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { Affiliation } from 'app/models/affiliation';
import { Person } from 'app/models/person';
import { Reference } from 'app/models/reference';
import { RelatedArticle } from 'app/models/related-article';
import { Keyword } from 'app/models/keyword';
import { AddObjectChange } from 'app/utils/history/add-object-change';

export function manuscriptEntityToJson<T>(object: T): JSONObject {
  return cloneDeepWith(object, (value) => {
    if (value instanceof EditorState) {
      return value.toJSON();
    }
  });
}

export function deserializeChanges(changesJson: JSONObject[]): Change[] {
  return changesJson.map((changeData) => {
    switch (changeData.type) {
      case 'prosemirror':
        return ProsemirrorChange.fromJSON(changeData);

      case 'rearranging':
        return RearrangingChange.fromJSON(changeData);

      case 'batch':
        return BatchChange.fromJSON(changeData);

      case 'update-object':
        return UpdateObjectChange.fromJSON(changeData);

      case 'delete-object':
        return DeleteObjectChange.fromJSON(changeData);

      case 'add-object':
        return AddObjectChange.fromJSON(changeData);

      default:
        throw new TypeError(
          `Cannot deserialize ${changeData.type} change. Change type invalid, JSON object malformed or handling not supported.`
        );
    }
  });
}

export function deserializeBackmatter(path: string, json: JSONObject): BackmatterEntity {
  if (path.indexOf('affiliations') >= 0) {
    return new Affiliation(json);
  }

  if (path.indexOf('authors') >= 0) {
    return new Person(json);
  }

  if (path.indexOf('references') >= 0) {
    return new Reference(json);
  }

  if (path.indexOf('relatedArticles') >= 0) {
    return new RelatedArticle(json);
  }

  if (path.match(/keywordGroups\.[^.]+\.keywords/)) {
    return new Keyword(json);
  }

  throw new Error(`deserialization of backmatter entity for  ${path} is not implemented or provided path is invalid`);
}

export function applyChangesFromServer(manuscript: Manuscript, changesJson: JSONObject[]): Manuscript {
  const allChanges = new BatchChange(deserializeChanges(changesJson));
  return allChanges.applyChange(manuscript);
}
