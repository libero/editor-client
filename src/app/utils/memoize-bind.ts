import { memoize } from 'lodash';
import objectHash from 'object-hash';

export const memoizeBind = memoize(
  (fn: Function, ...args: any[]) => {
    return fn.bind(null, ...args);
  },
  (fn: Function, firstArg: any) => {
    return objectHash({
      fn,
      firstArg
    });
  }
);
