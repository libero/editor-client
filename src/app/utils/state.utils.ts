import { EditorState } from 'prosemirror-state';
import { cloneDeepWith } from 'lodash';

import { Manuscript, ManuscriptDiff } from 'app/types/manuscript';

export interface LoadableState<T> {
  data: T | undefined;
  isLoading: boolean;
  error: Error | undefined;
}

export function getInitialLoadableState<T>(): LoadableState<T> {
  return {
    data: undefined,
    isLoading: false,
    error: undefined
  };
}

export function getLoadableStateProgress<T>(): LoadableState<T> {
  return {
    data: undefined,
    isLoading: true,
    error: undefined
  };
}

export function getLoadableStateSuccess<T>(data: T): LoadableState<T> {
  return {
    data,
    isLoading: false,
    error: undefined
  };
}

export function getLoadableStateError<T>(error: Error): LoadableState<T> {
  return {
    data: undefined,
    isLoading: false,
    error
  };
}

export interface ManuscriptHistory {
  past: ManuscriptDiff[];
  present: Manuscript;
  future: ManuscriptDiff[];
}

export function getInitialHistory(present: Manuscript): ManuscriptHistory {
  return { past: [], present, future: [] };
}

export function cloneManuscript(manuscript: Manuscript): Manuscript {
  const cloneCustomizer = (value): EditorState | undefined => (value instanceof EditorState ? value : undefined);
  return cloneDeepWith(manuscript, cloneCustomizer);
}
