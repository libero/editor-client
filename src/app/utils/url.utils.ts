/**
 * Extracts the value for the specified key from the supplied query parameters string
 *
 * @export
 * @param {string} params The query parameter string, e.g. '?name=joe&age=10'.
 * @param {string} key The key to find and return the value, e.g. 'name'.
 * @returns {(string | undefined)}
 */
export function getKeyFromQueryParams(params: string, key: string): string | undefined {
  let retVal = undefined;
  const regex = new RegExp(`A?${key}=([^&]+)&*`);
  const found = params.match(regex);
  if (found) {
    retVal = found[1];
  }
  return retVal;
}
