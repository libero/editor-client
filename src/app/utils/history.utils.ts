import { Transaction } from 'prosemirror-state';

import { ManuscriptHistory } from './state.utils';
import { ManuscriptDiff, ManuscriptDiffValues } from 'app/types/manuscript';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';

export function updateManuscriptState(
  state: ManuscriptHistory,
  propName: string,
  transaction: Transaction
): ManuscriptHistory {
  const change = new ProsemirrorChange(propName, transaction);

  if (transaction.docChanged) {
    return {
      past: [...state.past, change],
      present: change.applyChange(state.present),
      future: []
    } as ManuscriptHistory;
  } else {
    return {
      ...state,
      present: change.applyChange(state.present)
    };
  }
}

export function undoChange(state: ManuscriptHistory): ManuscriptHistory {
  const past = [...state.past];
  const change = past.pop();

  return {
    past,
    present: change.rollbackChange(state.present),
    future: [change, ...state.future]
  };
}

export function createDiff(changes: Record<string, ManuscriptDiffValues>): ManuscriptDiff {
  return {
    ...changes,
    _timestamp: Date.now()
  } as ManuscriptDiff;
}

export function redoChange(state: ManuscriptHistory): ManuscriptHistory {
  const future = [...state.future];
  const change = future.shift();

  return {
    past: [...state.past, change],
    present: change.applyChange(state.present),
    future: future
  };
}
