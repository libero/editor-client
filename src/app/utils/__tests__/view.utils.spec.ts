import { EditorState, TextSelection } from 'prosemirror-state';

import { hasParentNodeOf, objectsEqual, stringifyEditorState, getImageFileUpload } from 'app/utils/view.utils';
import { createBodyState } from 'app/models/body';

describe('view utils', () => {
  it('stringifies editor state content', () => {
    const editorState = givenEditorState();
    expect(stringifyEditorState(editorState)).toMatchSnapshot();
  });

  it('compares objects with editor state', () => {
    const obj1 = { value1: givenEditorState(), value2: 'Some value' };
    const obj2 = {
      value1: obj1.value1.apply(obj1.value1.tr.setSelection(TextSelection.create(obj1.value1.doc, 2, 3))),
      value2: 'Some value'
    };

    expect(objectsEqual(obj1, obj2)).toBeTruthy();
    expect(obj1.value1).not.toEqual(obj2.value1);
  });

  it('checks if node at position has specific parent', () => {
    const editorState = givenEditorState();
    const xrefPos = editorState.doc.resolve(57);
    expect(hasParentNodeOf(xrefPos, ['paragraph'])).toBeTruthy();
    expect(hasParentNodeOf(xrefPos, ['boxText'])).toBeTruthy();

    const pTextPos = editorState.doc.resolve(7);
    expect(hasParentNodeOf(pTextPos, ['paragraph'])).toBeTruthy();
    expect(hasParentNodeOf(pTextPos, ['boxText'])).toBeFalsy();
  });

  it('uploads an image', () => {
    const fileReader = new FileReaderMock();
    jest.spyOn(window, 'FileReader').mockImplementation(() => fileReader);

    const callback = jest.fn();
    jest.spyOn(document, 'createElement');
    getImageFileUpload(callback);
    const input = (document.createElement as jest.Mock).mock.results[0].value;
    const file = new File(['foo'], 'test.png', { type: 'image/png' });
    jest.spyOn(input.files.__proto__, 'item').mockReturnValue(file);

    input.dispatchEvent(new Event('change'));
    fileReader.onload({ target: { result: 'base 64 string' } });
    expect(callback).toBeCalledWith(expect.any(File));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});

function givenEditorState(): EditorState {
  const el = document.createElement('main-text');
  el.innerHTML = `
    <p>Example text</p>
    <p>Example text 2</p>
    <boxed-text>
      <p>
        <bold>Related research article</bold>
        <xref ref-type="bibr" rid="bib5">Harmon (2019)</xref>
      </p>
    </boxed-text>
  `;
  return createBodyState(el, '');
}

class FileReaderMock {
  DONE = FileReader.DONE;
  EMPTY = FileReader.EMPTY;
  LOADING = FileReader.LOADING;
  readyState = 0;
  error: FileReader['error'] = null;
  result: FileReader['result'] = null;
  abort = jest.fn();
  addEventListener = jest.fn();
  dispatchEvent = jest.fn();
  onabort = jest.fn();
  onerror = jest.fn();
  onload = jest.fn();
  onloadend = jest.fn();
  onloadprogress = jest.fn();
  onloadstart = jest.fn();
  onprogress = jest.fn();
  readAsArrayBuffer = jest.fn();
  readAsBinaryString = jest.fn();
  readAsDataURL = jest.fn();
  readAsText = jest.fn();
  removeEventListener = jest.fn();
}
