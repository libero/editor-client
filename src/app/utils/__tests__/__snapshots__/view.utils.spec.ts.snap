// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`view utils stringifies editor state content 1`] = `"<p class=\\"paragraph\\">Example text</p><p class=\\"paragraph\\">Example text 2</p><section class=\\"box-text\\"><p class=\\"paragraph\\"><strong>Related research article</strong> <a href=\\"#\\" class=\\"citation\\" data-cit-type=\\"reference\\" data-ref-id=\\"bib5\\" data-ref-text=\\"Harmon (2019)\\"></a></p></section>"`;
